tutorials/plugins/gdnative/files/cpp_example/SConstruct

```diff
@@ -1,48 +1,92 @@
-        if file.endswith('.cpp'):
-            sources.append(directory + '/' + file)
+# Local dependency paths, adapt them to your setup
+godot_headers_path = "godot-cpp/godot_headers/"
+cpp_bindings_path = "godot-cpp/"
+cpp_library = "libgodot-cpp"
 
-if platform == "osx":
-    env.Append(CCFLAGS = ['-g','-O3', '-arch', 'x86_64', '-std=c++14'])
-    env.Append(LINKFLAGS = ['-arch', 'x86_64'])
+# only support 64 at this time..
+bits = 64
 
-    final_lib_path = final_lib_path + 'osx/'
+# Updates the environment with the option variables.
+opts.Update(env)
 
-elif platform == "linux":
-    env.Append(CCFLAGS = ['-fPIC', '-g','-O3', '-std=c++14'])
+# Process some arguments
+if env['use_llvm']:
+    env['CXX'] = 'clang++'
 
-    final_lib_path = final_lib_path + 'x11/'
+if env['p'] != 'not set':
+    env['platform'] = env['p']
 
-elif platform == "windows":
-    if target == "debug":
-        env.Append(CCFLAGS = ['-EHsc', '-D_DEBUG', '-MDd'])
+# Check our platform specifics
+if env['platform'] == "osx":
+    env['target_path'] += 'osx/'
+    cpp_library += '.osx'
+    if env['target'] in ('debug', 'd'):
+        env.Append(CCFLAGS = ['-g','-O2', '-arch', 'x86_64'])
+        env.Append(LINKFLAGS = ['-arch', 'x86_64'])
     else:
-        env.Append(CCFLAGS = ['-O2', '-EHsc', '-DNDEBUG', '-MD'])
+        env.Append(CCFLAGS = ['-g','-O3', '-arch', 'x86_64'])
+        env.Append(LINKFLAGS = ['-arch', 'x86_64'])
 
-    final_lib_path = final_lib_path + 'win' + str(bits) + '/'
+elif env['platform'] in ('x11', 'linux'):
+    env['target_path'] += 'x11/'
+    cpp_library += '.linux'
+    if env['target'] in ('debug', 'd'):
+        env.Append(CCFLAGS = ['-fPIC', '-g3','-Og', '-std=c++17'])
+    else:
+        env.Append(CCFLAGS = ['-fPIC', '-g','-O3', '-std=c++17'])
+
+elif env['platform'] == "windows":
+    env['target_path'] += 'win64/'
+    cpp_library += '.windows'
+    if not env['use_llvm']:
+        # This makes sure to keep the session environment variables on windows,
+        # that way you can run scons in a vs 2017 prompt and it will find all the required tools
+        env.Append(ENV = os.environ)
+
+        env.Append(CCFLAGS = ['-DWIN32', '-D_WIN32', '-D_WINDOWS', '-W3', '-GR', '-D_CRT_SECURE_NO_WARNINGS'])
+        if env['target'] in ('debug', 'd'):
+            env.Append(CCFLAGS = ['-EHsc', '-D_DEBUG', '-MDd'])
+        else:
+            env.Append(CCFLAGS = ['-O2', '-EHsc', '-DNDEBUG', '-MD'])
+    # untested
+    else:
+        if env['target'] in ('debug', 'd'):
+            env.Append(CCFLAGS = ['-fPIC', '-g3','-Og', '-std=c++17'])
+        else:
+            env.Append(CCFLAGS = ['-fPIC', '-g','-O3', '-std=c++17'])
+
+if env['target'] in ('debug', 'd'):
+    cpp_library += '.debug'
+else:
+    cpp_library += '.release'
+
+cpp_library += '.' + str(bits)
+
+env.Append(CPPPATH=['.', 'src/', godot_headers_path, cpp_bindings_path + 'include/', cpp_bindings_path + 'include/core/', cpp_bindings_path + 'include/gen/'])
```

Comment:
If `src/` is meant to be the user-defined location where they put their sources, I'd suggest to split this in two lines: one with the include paths that *must* be used to link against the GDNative headers, and one line with the include paths that the user would want to adapt to their needs.

That is, if we want this example to serve as a reference buildsystem for GDNative projects (which is IMO needed as many seem to struggle with it). If it's a reference, there should be a clear visibility of what is the base buildsystem setup that should likely be kept as is, and what are the parts where the user will want to tweak things (e.g. the C++17 flags above, are those necessary, or just an example that they can be used if needed?).

Reply:
Overall it's a good SConstruct example though. Eventually we should look into moving parts of it (and Godot's) to Python modules that could be shared between Godot and GDNative C/C++. Especially all the stuff regarding building the binary name with platform, target and arch info, but we should probably wait for #24030 before doing that.

Reply:
I think its a pretty decent reference, Will has made a nice writeup but his SConstruct file has much more stuff in it. I kinda like having some things by convention so people just hit the ground running.

Do agree with splitting it in two lines, and looking forward to see what comes out of splitting the SConstruct up more.